lib/services/role.go

```diff
@@ -172,6 +173,12 @@ type Role interface {
 	// Equals returns true if the roles are equal. Roles are equal if options and
 	// conditions match.
 	Equals(other Role) bool
+	// Clone returns a deep copy of a role.
+	Clone() Role
+	// ApplyContext makes a deep copy of the role and applies the passed in
+	// context to any variables in the list of logins and returns the deep copy
+	// of the role.
+	ApplyContext(map[string][]string) Role
```

Comment:
When reading this code I wished repeatedly for a context to be declared as its own type:

```
type RoleContext map[string][]string
```
It would make the code more readable and you could attach some helper methods to it as well.

Is using a native map more inconvenient?